coordinator/tests/integration/test_service_mesh.py

```diff
@@ -0,0 +1,251 @@
+    )
+
+
+@pytest.mark.setup
+def test_build_deploy_tester_grpc(juju: Juju):
+    path = "./tests/integration/tester-grpc/"
+    charm = pack(path).absolute()
+    resources = get_resources(path)
+    juju.deploy(
+        charm,
+        TESTER_GRPC_APP_NAME,
+        resources=resources,
+        num_units=3,
+    )
+
+
+@pytest.mark.setup
+def test_deploy_monolithic_cluster(juju: Juju, tempo_charm: Path):
+    deploy_monolithic_cluster(juju)
+
+
+@pytest.mark.setup
+def test_scale_up_tempo(juju: Juju):
+    juju.add_unit(TEMPO_APP, num_units=2)
+    juju.wait(
+        lambda status: jubilant.all_active(status, TEMPO_APP, WORKER_APP), timeout=1000
+    )
+
+
+@pytest.mark.setup
+def test_relate(juju: Juju):
+    juju.integrate(TEMPO_APP + ":tracing", TESTER_APP_NAME + ":tracing")
+    juju.integrate(TEMPO_APP + ":tracing", TESTER_GRPC_APP_NAME + ":tracing")
+    # Add tempo_api relation for testing API access
+    juju.integrate(TEMPO_APP + ":tempo-api", TESTER_APP_NAME + ":tempo-api")
+
+    juju.wait(
+        lambda status: jubilant.all_active(
+            status, TEMPO_APP, WORKER_APP, TESTER_APP_NAME, TESTER_GRPC_APP_NAME
+        ),
+        timeout=1000,
+    )
+
+
+@pytest.mark.setup
+def test_deploy_istio(juju: Juju):
+    # Deploy Istio core
+    deploy_istio(juju)
+    juju.wait(
+        lambda status: jubilant.all_active(status, ISTIO_APP),
+        timeout=1000,
+    )
+
+
+@pytest.mark.setup
+def test_deploy_istio_beacon(juju: Juju):
+    # Deploy Istio beacon
+    deploy_istio_beacon(juju)
+    # Configure istio-beacon to enable mesh for the model
+    juju.config(ISTIO_BEACON_APP, {"model-on-mesh": "true"})
+    juju.wait(
+        lambda status: jubilant.all_active(status, ISTIO_BEACON_APP),
+        timeout=1000,
+    )
+    # Relate istio-beacon with tempo for service mesh
+    juju.integrate(ISTIO_BEACON_APP + ":service-mesh", TEMPO_APP + ":service-mesh")
+
+
+@pytest.mark.setup
+def test_add_mesh_policies_and_restart_relations(juju: Juju):
+    import subprocess
+    import tempfile
+
+    # Create the tester-grpc peer policy - keep this manual here so the tester charm doesnt need a service mesh integration.
+    # Get the model name from juju which is the namespace in k8s
+    model_name = juju.model
+    policy_content = f"""apiVersion: security.istio.io/v1
+kind: AuthorizationPolicy
+metadata:
+  name: allow-tester-grpc-peer
+  namespace: {model_name}
+  labels:
+    app.juju.is/created-by: tempo
+spec:
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: tester-grpc
+  rules:
+  - from:
+    - source:
+        principals:
+        - cluster.local/ns/{model_name}/sa/tester-grpc
+    to:
+    - operation:
+        ports:
+        - "3000"  # Tester service port
+"""
+
+    with tempfile.NamedTemporaryFile(mode="w", suffix=".yaml", delete=False) as f:
+        f.write(policy_content)
+        f.flush()
+        try:
+            result = subprocess.run(
+                ["kubectl", "apply", "-f", f.name],
```

Comment:
this assumes that kubectl is present on the host (and configured to point to the right cluster). Can we use lightkube instead? Or verify this assumption and raise an exception earlier on?